core/math/a_star_grid_2d.h

```diff
@@ -0,0 +1,207 @@
+/* Copyright (c) 2014-2022 Godot Engine contributors (cf. AUTHORS.md).   */
+/*                                                                       */
+/* Permission is hereby granted, free of charge, to any person obtaining */
+/* a copy of this software and associated documentation files (the       */
+/* "Software"), to deal in the Software without restriction, including   */
+/* without limitation the rights to use, copy, modify, merge, publish,   */
+/* distribute, sublicense, and/or sell copies of the Software, and to    */
+/* permit persons to whom the Software is furnished to do so, subject to */
+/* the following conditions:                                             */
+/*                                                                       */
+/* The above copyright notice and this permission notice shall be        */
+/* included in all copies or substantial portions of the Software.       */
+/*                                                                       */
+/* THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,       */
+/* EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF    */
+/* MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.*/
+/* IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY  */
+/* CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,  */
+/* TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE     */
+/* SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.                */
+/*************************************************************************/
+
+#ifndef A_STAR_GRID_2D_H
+#define A_STAR_GRID_2D_H
+
+#include "core/object/gdvirtual.gen.inc"
+#include "core/object/ref_counted.h"
+#include "core/object/script_language.h"
+#include "core/templates/list.h"
+#include "core/templates/oa_hash_map.h"
+
+class AStarGrid2D : public RefCounted {
+	GDCLASS(AStarGrid2D, RefCounted);
+
+public:
+	enum DiagonalMode {
+		DIAGONAL_MODE_ALWAYS,
+		DIAGONAL_MODE_NEVER,
+		DIAGONAL_MODE_AT_LEAST_ONE_WALKABLE,
+		DIAGONAL_MODE_ONLY_IF_NO_OBSTACLES,
+		DIAGONAL_MODE_MAX,
+	};
+
+	enum Heuristic {
+		HEURISTIC_EUCLIDEAN,
+		HEURISTIC_MANHATTAN,
+		HEURISTIC_OCTILE,
+		HEURISTIC_CHEBYSHEV,
+		HEURISTIC_MAX,
+	};
+
+private:
+	int64_t width = 0;
+	int64_t height = 0;
+
+	bool jumping_enabled = false;
+	DiagonalMode diagonal_mode = DIAGONAL_MODE_ALWAYS;
+	Heuristic default_heuristic = HEURISTIC_EUCLIDEAN;
+
+	enum Dir {
+		DIR_LEFT,
+		DIR_RIGHT,
+		DIR_TOP,
+		DIR_TOP_LEFT,
+		DIR_TOP_RIGHT,
+		DIR_BOTTOM,
+		DIR_BOTTOM_LEFT,
+		DIR_BOTTOM_RIGHT,
+	};
+
+	struct Point {
+		int64_t id = 0;
+
+		_FORCE_INLINE_ int64_t get_x(int64_t p_grid_width) { return id % p_grid_width; }
+		_FORCE_INLINE_ int64_t get_y(int64_t p_grid_width) { return id / p_grid_width; }
+
+		real_t weight_scale = 0;
+		bool enabled = true;
+		Vector2 pos;
+
+		Point *neighbours[8] = { nullptr };
```

Comment:
If this is a grid, would it not make more sense that the neighbours are just computed by code? You should not need pointers and this would save memory.

Reply:
Fixed